lib/services/saml.go

```diff
@@ -0,0 +1,113 @@
+package services
+
+import (
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/xml"
+	"strings"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	saml2 "github.com/russellhaering/gosaml2"
+	"github.com/russellhaering/gosaml2/types"
+	dsig "github.com/russellhaering/goxmldsig"
+)
+
+// GetAttributeNames returns a list of claim names from the claim values
+func GetAttributeNames(attributes map[string]types.Attribute) []string {
+	var out []string
+	for _, attr := range attributes {
+		out = append(out, attr.Name)
+	}
+	return out
+}
+
+// SAMLAssertionsToTraits converts saml assertions to traits
+func SAMLAssertionsToTraits(assertions saml2.AssertionInfo) map[string][]string {
+	traits := make(map[string][]string, len(assertions.Values))
+	for _, assr := range assertions.Values {
+		vals := make([]string, 0, len(assr.Values))
+		for _, value := range assr.Values {
+			vals = append(vals, value.Value)
+		}
+		traits[assr.Name] = vals
+	}
+	return traits
+}
+
+// GetSAMLServiceProvider gets the SAMLConnector's service provider
+func GetSAMLServiceProvider(sc SAMLConnector, clock clockwork.Clock) (*saml2.SAMLServiceProvider, error) {
+	certStore := dsig.MemoryX509CertificateStore{
+		Roots: []*x509.Certificate{},
+	}
+
+	metadata := &types.EntityDescriptor{}
+	err := xml.Unmarshal([]byte(sc.GetEntityDescriptor()), metadata)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to parse entity_descriptor")
+	}
+
+	for _, kd := range metadata.IDPSSODescriptor.KeyDescriptors {
+		for _, samlCert := range kd.KeyInfo.X509Data.X509Certificates {
+			certData, err := base64.StdEncoding.DecodeString(strings.TrimSpace(samlCert.Data))
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			cert, err := x509.ParseCertificate(certData)
+			if err != nil {
+				return nil, trace.Wrap(err, "failed to parse certificate in metadata")
+			}
+			certStore.Roots = append(certStore.Roots, cert)
+		}
+	}
```

Comment:
IIUC, the old `GetServiceProvider` allows `sc.Spec.EntityDescriptor` to be nil.
Here, it's always parsed unconditionally.